wcfsetup/install/files/ts/WoltLabSuite/Core/Ui/ItemList/Static.ts

```diff
@@ -0,0 +1,443 @@
+/**
+ * Flexible UI element featuring both a list of items and an input field.
+ *
+ * @author  Alexander Ebert, Matthias Schmidt
+ * @copyright  2001-2019 WoltLab GmbH
+ * @license  GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module  WoltLabSuite/Core/Ui/ItemList/Static
+ */
+
+import * as Core from "../../Core";
+import * as DomTraverse from "../../Dom/Traverse";
+import * as Language from "../../Language";
+import UiDropdownSimple from "../Dropdown/Simple";
+
+export type CallbackChange = (elementId: string, values: ItemData[]) => void;
+export type CallbackSubmit = (form: HTMLFormElement, values: ItemData[]) => void;
+
+export interface ItemListStaticOptions {
+  maxItems: number;
+  maxLength: number;
+  isCSV: boolean;
+  callbackChange: CallbackChange | null;
+  callbackSubmit: CallbackSubmit | null;
+  submitFieldName: string;
+}
+
+type ItemListInputElement = HTMLInputElement | HTMLTextAreaElement;
+
+export interface ItemData {
+  objectId: number;
+  value: string;
+  type?: string;
+}
+
+type PlainValue = string;
+
+type ItemDataOrPlainValue = ItemData | PlainValue;
+
+interface UiData {
+  element: HTMLInputElement | HTMLTextAreaElement;
+  list: HTMLOListElement;
+  shadow?: HTMLInputElement;
+  values: string[];
+}
+
+interface ElementData {
+  dropdownMenu: HTMLElement | null;
+  element: ItemListInputElement;
+  list: HTMLOListElement;
+  listItem: HTMLElement;
+  options: ItemListStaticOptions;
+  shadow?: HTMLInputElement;
+}
+
+const _data = new Map<string, ElementData>();
+
+/**
+ * Creates the DOM structure for target element. If `element` is a `<textarea>`
+ * it will be automatically replaced with an `<input>` element.
+ */
+function createUI(element: ItemListInputElement, options: ItemListStaticOptions): UiData {
+  const list = document.createElement("ol");
+  list.className = "inputItemList" + (element.disabled ? " disabled" : "");
+  list.dataset.elementId = element.id;
+  list.addEventListener("click", (event) => {
+    if (event.target === list) {
+      element.focus();
+    }
+  });
+
+  const listItem = document.createElement("li");
+  listItem.className = "input";
+  list.appendChild(listItem);
+
+  element.addEventListener("keydown", (ev: KeyboardEvent) => keyDown(ev));
```

Comment:
Why do you need to add the type here? TypeScript *should* be able to automatically derive it.

Reply:
For some reason PhpStorm is unable to determine the type because of the type for the union type. If it move the type union into the argument list, it works, but using the `ItemListInputElement` type bricks it.

It's a bit verbose, but prevents my IDE from lighting up like a christmas tree.

Reply:
That's a good rationale.